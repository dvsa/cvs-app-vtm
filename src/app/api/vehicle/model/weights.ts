import { AxleBrakeProperties } from './axleBrakeProperties';
import { AxleTyreProperties } from './axleTyreProperties';
import { AxleWeightProperties } from './axleWeightProperties';

export interface Weights { 
    /**
     * Used for all vehicle types - PSV, HGV and TRL
     */
    axleNumber?: number;
    /**
     * Used for all vehicle types - PSV, HGV and TRL. Optional for HGV
     */
    parkingBrakeMrk?: boolean;
    weights?: AxleWeightProperties;
    tyres?: AxleTyreProperties;
    brakes?: AxleBrakeProperties;
}