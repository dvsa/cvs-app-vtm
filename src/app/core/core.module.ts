import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GlobalErrorComponent } from './components/global-error/global-error.component';
import { SpinnerComponent } from './components/spinner/spinner.component';
import { FooterComponent } from './components/footer/footer.component';
import { HeaderComponent } from './components/header/header.component';
import { RouterModule } from '@angular/router';
import { BreadcrumbsComponent } from './components/breadcrumbs/breadcrumbs.component';
import { PageNotFoundComponent } from './components/page-not-found/page-not-found.component';
import { ServerErrorComponent } from './components/server-error/server-error.component';

@NgModule({
  declarations: [FooterComponent, HeaderComponent, GlobalErrorComponent, SpinnerComponent, BreadcrumbsComponent, PageNotFoundComponent, ServerErrorComponent],
  imports: [CommonModule, RouterModule],
  exports: [FooterComponent, HeaderComponent, GlobalErrorComponent, SpinnerComponent, BreadcrumbsComponent]
})
export class CoreModule {}
