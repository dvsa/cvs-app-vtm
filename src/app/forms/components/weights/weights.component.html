<ng-container *ngIf="isEditable">
  <form [formGroup]="form">
    <table class="govuk-table">
      <caption class="govuk-table__caption govuk-table__caption--m">
        Axle weights (kg)
      </caption>

      <thead class="govuk-table__head">
        <tr class="govuk-table__row">
          <th scope="col" class="govuk-table__header"></th>
          <th scope="col" class="govuk-table__header">Kerb</th>
          <th scope="col" class="govuk-table__header">Laden</th>
          <th scope="col" class="govuk-table__header">GB max</th>
          <th scope="col" class="govuk-table__header">Design</th>
        </tr>
      </thead>
      <tbody class="govuk-table__body">
        <ng-container formArrayName="axles">
          <tr class="govuk-table__row" *ngFor="let axle of axlesFormArray.controls; let i = index">
            <ng-container [formGroupName]="i">
              <th scope="row" class="govuk-table__header">Axle {{ axle.value.axleNumber }}</th>
              <ng-container formGroupName="weights">
                <td class="govuk-table__cell">
                  <app-text-input name="foo" formControlName="kerbWeight"></app-text-input>
                </td>
                <td class="govuk-table__cell"><app-text-input name="foo" formControlName="ladenWeight"></app-text-input></td>
                <td class="govuk-table__cell"><app-text-input name="foo" formControlName="gbWeight"></app-text-input></td>
                <td class="govuk-table__cell"><app-text-input name="foo" formControlName="designWeight"></app-text-input></td>
              </ng-container>
            </ng-container>
          </tr>
        </ng-container>
        <tr class="govuk-table__row">
          <th scope="row" class="govuk-table__header">GROSS</th>
          <td class="govuk-table__cell"><app-text-input name="foo" formControlName="grossKerbWeight"></app-text-input></td>
          <td class="govuk-table__cell"><app-text-input name="foo" formControlName="grossLadenWeight"></app-text-input></td>
          <td class="govuk-table__cell"><app-text-input name="foo" formControlName="grossGbWeight"></app-text-input></td>
          <td class="govuk-table__cell"><app-text-input name="foo" formControlName="grossDesignWeight"></app-text-input></td>
        </tr>
        <tr class="govuk-table__row">
          <th scope="row" class="govuk-table__header">TRAIN</th>
          <td class="govuk-table__cell"></td>
          <td class="govuk-table__cell"></td>
          <td class="govuk-table__cell"><app-text-input name="foo" formControlName="maxTrainGbWeight"></app-text-input></td>
          <td class="govuk-table__cell"><app-text-input name="foo" formControlName="trainDesignWeight"></app-text-input></td>
        </tr>
      </tbody>
    </table>

    <app-text-input name="foo" formControlName="unladenWeight"></app-text-input>
  </form>
</ng-container>

<ng-container *ngIf="!isEditable">
  <table class="govuk-table">
    <caption class="govuk-table__caption govuk-table__caption--m">
      Axle weights (kg)
    </caption>

    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th scope="col" class="govuk-table__header"></th>
        <th scope="col" class="govuk-table__header">Kerb</th>
        <th scope="col" class="govuk-table__header">Laden</th>
        <th scope="col" class="govuk-table__header">GB max</th>
        <th scope="col" class="govuk-table__header">Design</th>
      </tr>
    </thead>
    <tbody class="govuk-table__body">
      <tr class="govuk-table__row" *ngFor="let axle of data.axles; let i = index">
        <th scope="row" class="govuk-table__header">Axle {{ axle.axleNumber }}</th>
        <td class="govuk-table__cell">{{ axle.weights?.kerbWeight }}</td>
        <td class="govuk-table__cell">{{ axle.weights?.ladenWeight }}</td>
        <td class="govuk-table__cell">{{ axle.weights?.gbWeight }}</td>
        <td class="govuk-table__cell">{{ axle.weights?.designWeight }}</td>
      </tr>
      <tr class="govuk-table__row">
        <th scope="row" class="govuk-table__header">Gross</th>
        <td class="govuk-table__cell">{{ data.grossKerbWeight }}</td>
        <td class="govuk-table__cell">{{ data.grossLadenWeight }}</td>
        <td class="govuk-table__cell">{{ data.grossGbWeight }}</td>
        <td class="govuk-table__cell">{{ data.grossDesignWeight }}</td>
      </tr>
      <tr class="govuk-table__row">
        <th scope="row" class="govuk-table__header">Train</th>
        <td class="govuk-table__cell"></td>
        <td class="govuk-table__cell"></td>
        <td class="govuk-table__cell">{{ data.maxTrainGbWeight }}</td>
        <td class="govuk-table__cell">{{ data.trainDesignWeight }}</td>
      </tr>
    </tbody>
  </table>
  <!-- TODO: come here and sort this-->
  {{ data.unladenWeight }}
</ng-container>
