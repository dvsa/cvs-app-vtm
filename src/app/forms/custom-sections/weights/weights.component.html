<form [formGroup]="form">
  <table class="govuk-table">
    <caption class="govuk-table__caption govuk-table__caption--m">
      Axle weights (kg)
    </caption>

    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th scope="col" class="govuk-table__header"></th>

        <ng-container *ngIf="isPsv; else hgvAndTrlTh">
          <th scope="col" class="govuk-table__header">Kerb</th>
          <th scope="col" class="govuk-table__header">Laden</th>
          <th scope="col" class="govuk-table__header">GB max</th>
          <th scope="col" class="govuk-table__header">Design</th>
          <th *ngIf="isEditing" scope="col" class="govuk-table__header"></th>
        </ng-container>

        <ng-template #hgvAndTrlTh>
          <th scope="col" class="govuk-table__header">GB</th>
          <th scope="col" class="govuk-table__header">EEC</th>
          <th scope="col" class="govuk-table__header">Design</th>
          <th *ngIf="isEditing" scope="col" class="govuk-table__header"></th>
        </ng-template>
      </tr>
    </thead>

    <tbody class="govuk-table__body">
      <tr *ngFor="let axle of axles.controls; let i = index" class="govuk-table__row" formArrayName="axles">
        <ng-container [formGroupName]="i">
          <th scope="row" class="govuk-table__header">Axle {{ axle.value.axleNumber }}</th>

          <ng-container formGroupName="weights">
            <ng-container *ngIf="isPsv">
              <ng-template [ngTemplateOutlet]="axleTdNumberInput" [ngTemplateOutletContext]="{ name: 'kerbWeight' }"></ng-template>

              <ng-template [ngTemplateOutlet]="axleTdNumberInput" [ngTemplateOutletContext]="{ name: 'ladenWeight' }"></ng-template>
            </ng-container>

            <ng-template [ngTemplateOutlet]="axleTdNumberInput" [ngTemplateOutletContext]="{ name: 'gbWeight' }"></ng-template>

            <ng-template *ngIf="!isPsv" [ngTemplateOutlet]="axleTdNumberInput" [ngTemplateOutletContext]="{ name: 'eecWeight' }"></ng-template>

            <ng-template [ngTemplateOutlet]="axleTdNumberInput" [ngTemplateOutletContext]="{ name: 'designWeight' }"></ng-template>

            <td *ngIf="isEditing" class="govuk-table__cell">
              <a class="axleButton" (click)="removeAxle(i)">Remove</a>
            </td>

            <ng-template #axleTdNumberInput let-name="name">
              <td class="govuk-table__cell">
                <app-switchable-input
                  [form]="getAxleWeights(i)"
                  [type]="types.NUMBER"
                  [name]="name"
                  [idExtension]="axle.value.axleNumber"
                  [isEditing]="isEditing"
                ></app-switchable-input>
              </td>
            </ng-template>
          </ng-container>
        </ng-container>
      </tr>

      <ng-container *ngIf="isPsv; else hgvAndTrlWeights">
        <tr class="govuk-table__row">
          <th scope="row" class="govuk-table__header">Gross</th>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossKerbWeight' }"></ng-template>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossLadenWeight' }"></ng-template>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossGbWeight' }"></ng-template>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossDesignWeight' }"></ng-template>
          <td *ngIf="isEditing" class="govuk-table__cell"></td>
        </tr>

        <tr class="govuk-table__row">
          <th scope="row" class="govuk-table__header">Train</th>
          <td class="govuk-table__cell"></td>
          <td class="govuk-table__cell"></td>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'maxTrainGbWeight' }"></ng-template>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'trainDesignWeight' }"></ng-template>
          <td *ngIf="isEditing" class="govuk-table__cell"></td>
        </tr>
      </ng-container>

      <ng-template #hgvAndTrlWeights>
        <tr class="govuk-table__row">
          <th scope="row" class="govuk-table__header">Gross</th>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossGbWeight' }"></ng-template>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossEecWeight' }"></ng-template>
          <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'grossDesignWeight' }"></ng-template>
          <td *ngIf="isEditing" class="govuk-table__cell"></td>
        </tr>

        <ng-container *ngIf="isHgv">
          <tr class="govuk-table__row">
            <th scope="row" class="govuk-table__header">Train</th>
            <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'trainGbWeight' }"></ng-template>
            <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'trainEecWeight' }"></ng-template>
            <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'trainDesignWeight' }"></ng-template>
            <td *ngIf="isEditing" class="govuk-table__cell"></td>
          </tr>

          <tr class="govuk-table__row">
            <th scope="row" class="govuk-table__header">Max Train</th>
            <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'maxTrainGbWeight' }"></ng-template>
            <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'maxTrainEecWeight' }"></ng-template>
            <ng-template [ngTemplateOutlet]="tdNumberInput" [ngTemplateOutletContext]="{ name: 'maxTrainDesignWeight' }"></ng-template>
            <td *ngIf="isEditing" class="govuk-table__cell"></td>
          </tr>
        </ng-container>
      </ng-template>

      <tr *ngIf="isEditing">
        <a class="axleButton" role="button" id="weightsAddAxle" (click)="addAxle()">Add Axle</a>
        <p *ngIf="isError" class="govuk-error-message">{{ errorMessage }}</p>
      </tr>

      <ng-container *ngIf="isPsv">
        <br />

        <app-switchable-input
          [form]="form"
          [type]="types.NUMBER"
          name="unladenWeight"
          label="Unladen Weight"
          [isEditing]="isEditing"
          suffix="kg"
        ></app-switchable-input>
      </ng-container>

      <ng-template #tdNumberInput let-name="name">
        <td class="govuk-table__cell">
          <app-switchable-input [form]="form" [type]="types.NUMBER" [name]="name" [isEditing]="isEditing"></app-switchable-input>
        </td>
      </ng-template>
    </tbody>
  </table>
</form>
