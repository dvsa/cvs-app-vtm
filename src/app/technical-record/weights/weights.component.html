<table class="govuk-table table-details">
  <thead class="govuk-table__head">
  <tr class="govuk-table__row">
    <th scope="col" class="govuk-table__header"></th>
    <th scope="col" class="govuk-table__header">GB weight (kg)</th>
    <th scope="col" class="govuk-table__header">Design weight (kg)</th>
    <th class="govuk-table__cell"></th>
    <th class="govuk-table__cell"></th>
    <th class="govuk-table__cell"></th>
  </tr>
  </thead>
  <tbody class="govuk-table__body">
  <tr class="govuk-table__row">
    <th class="govuk-table__cell govuk-!-font-weight-bold">Gross</th>
    <td id="test-grossGbWeight" class="govuk-table__cell">
      {{ activeRecord.grossGbWeight | DefaultNullOrEmpty }}
    </td>
    <td id="test-grossDesignWeight" class="govuk-table__cell">
      {{ activeRecord.grossDesignWeight | DefaultNullOrEmpty }}
    </td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
  </tr>
  <tr class="govuk-table__row" *ngIf="activeRecord.vehicleType == 'hgv'">
    <th class="govuk-table__cell govuk-!-font-weight-bold">Train</th>
    <td id="test-trainGbWeight" class="govuk-table__cell">
      {{ activeRecord.trainGbWeight | DefaultNullOrEmpty }}
    </td>
    <td id="test-trainDesignWeight" class="govuk-table__cell">
      {{ activeRecord.trainDesignWeight | DefaultNullOrEmpty }}
    </td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
  </tr>
  <tr class="govuk-table__row" *ngIf="activeRecord.vehicleType == 'hgv'">
    <th class="govuk-table__cell govuk-!-font-weight-bold">Max train</th>
    <td id="test-maxTrainGbWeight" class="govuk-table__cell">
      {{ activeRecord.maxTrainGbWeight | DefaultNullOrEmpty }}
    </td>
    <td id="test-minTrainGbWeight" class="govuk-table__cell">
      {{ activeRecord.maxTrainDesignWeight | DefaultNullOrEmpty }}
    </td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
  </tr>
  <tr class="govuk-table__row" *ngFor="let axelNo of activeRecord.axles">
    <th class="govuk-table__cell govuk-!-font-weight-bold">Axle
      {{ axelNo.axleNumber | DefaultNullOrEmpty }}</th>
    <td id="{{ 'test-gbWeight-' + axelNo.axleNumber }}" class="govuk-table__cell">
      {{ axelNo.weights ? (axelNo.weights.gbWeight | DefaultNullOrEmpty) : '-' }}</td>
    <td id="{{ 'test-designWeight-' + axelNo.axleNumber }}" class="govuk-table__cell">
      {{ axelNo.weights ? (axelNo.weights.designWeight | DefaultNullOrEmpty) : '-' }}</td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
    <td class="govuk-table__cell"></td>
  </tr>
  </tbody>
</table>
