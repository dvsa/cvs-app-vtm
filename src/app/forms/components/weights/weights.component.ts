import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { FormArray } from '@angular/forms';
import { DynamicFormService } from '@forms/services/dynamic-form.service';
import { CustomFormGroup, FormNode } from '@forms/services/dynamic-form.types';
import { Axle, TechRecordModel } from '@models/vehicle-tech-record.model';
import { debounceTime, Subscription } from 'rxjs';

@Component({
  selector: 'app-weights',
  templateUrl: './weights.component.html'
})
export class WeightsComponent implements OnInit, OnDestroy {
  @Input() isEditable = false;
  @Input() template: FormNode | undefined;
  @Input() data: Partial<TechRecordModel> = {};

  @Output() formChange = new EventEmitter();

  public form!: CustomFormGroup;
  private _formSubscription = new Subscription();

  constructor(private dfs: DynamicFormService) {}

  ngOnInit(): void {
    this.form = this.dfs.createForm(this.template!, this.data) as CustomFormGroup;
    this._formSubscription = this.form.cleanValueChanges.pipe(debounceTime(400)).subscribe(event => {
      this.formChange.emit(event);
    });
  }

  ngOnDestroy(): void {
    this._formSubscription.unsubscribe();
  }

  get axles(): Axle[] | undefined {
    return this.data.axles;
  }

  get axlesFormArray() {
    return this.form.get(['axles']) as FormArray;
  }
}
