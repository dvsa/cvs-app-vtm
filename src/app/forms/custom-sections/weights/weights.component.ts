import { Component, EventEmitter, Input, OnChanges, OnDestroy, OnInit, Output, SimpleChanges } from '@angular/core';
import { TechRecordType } from '@dvsa/cvs-type-definitions/types/v3/tech-record/tech-record-vehicle-type';
import { DynamicFormService } from '@forms/services/dynamic-form.service';
import { CustomFormArray, CustomFormGroup, FormNode, FormNodeEditTypes } from '@forms/services/dynamic-form.types';
import { HgvWeight } from '@forms/templates/hgv/hgv-weight.template';
import { PsvWeightsTemplate } from '@forms/templates/psv/psv-weight.template';
import { TrlWeight } from '@forms/templates/trl/trl-weight.template';
import { Axle, VehicleTypes } from '@models/vehicle-tech-record.model';
import { Store } from '@ngrx/store';
import { addAxle, removeAxle, updateBrakeForces } from '@store/technical-records';
import { TechnicalRecordServiceState } from '@store/technical-records/reducers/technical-record-service.reducer';
import { debounceTime, Subscription } from 'rxjs';

@Component({
  selector: 'app-weights[vehicleTechRecord]',
  templateUrl: './weights.component.html',
  styleUrls: ['./weights.component.scss']
})
export class WeightsComponent implements OnInit, OnDestroy, OnChanges {
  @Input() vehicleTechRecord!: TechRecordType<'psv'> | TechRecordType<'trl'> | TechRecordType<'hgv'>;
  @Input() isEditing = false;

  @Output() formChange = new EventEmitter();

  public form!: CustomFormGroup;
  private _formSubscription = new Subscription();
  public isError: boolean = false;
  public errorMessage?: string;
  private ladenWeightOverride: boolean = false;

  constructor(public dynamicFormsService: DynamicFormService, private store: Store<TechnicalRecordServiceState>) {}

  ngOnInit(): void {
    this.form = this.dynamicFormsService.createForm(this.template, this.vehicleTechRecord) as CustomFormGroup;

    const grossLadenWeightChanges = this.form.get('techRecord_grossLadenWeight')?.valueChanges.subscribe(() => {
      this.ladenWeightOverride = true;
    });
    if (grossLadenWeightChanges) {
      this._formSubscription.add(grossLadenWeightChanges);
    }

    this._formSubscription.add(
      this.form.cleanValueChanges.pipe(debounceTime(400)).subscribe((event: any) => {
        if (this.ladenWeightOverride) return;

        const {
          techRecord_seatsUpperDeck,
          techRecord_seatsLowerDeck,
          techRecord_manufactureYear,
          techRecord_grossKerbWeight,
          techRecord_grossLadenWeight
        } = event || {};

        const shouldRecalculate =
          techRecord_seatsUpperDeck !== undefined ||
          techRecord_seatsLowerDeck !== undefined ||
          techRecord_manufactureYear !== undefined ||
          techRecord_grossKerbWeight !== undefined;

        if (shouldRecalculate) {
          this.ladenWeightOverride = false;
        }

        if (event?.techRecord_axles) {
          event.techRecord_axles = (event.techRecord_axles as Axle[]).filter(axle => !!axle?.axleNumber);
        }

        if (this.isPsv && !this.ladenWeightOverride && shouldRecalculate) {
          const calculatedWeight = this.calculateGrossLadenWeight();
          event.techRecord_grossLadenWeight = calculatedWeight;
          this.form.get('techRecord_grossLadenWeight')?.setValue(calculatedWeight, { emitEvent: false });
        }
        this.formChange.emit(event);

        if (techRecord_grossLadenWeight || techRecord_grossKerbWeight) {
          this.store.dispatch(updateBrakeForces({ grossLadenWeight: techRecord_grossLadenWeight, grossKerbWeight: techRecord_grossKerbWeight }));
        }
      })
    );
  }

  ngOnChanges(changes: SimpleChanges): void {
    const { vehicleTechRecord } = changes;
    if (this.form && vehicleTechRecord && !this.ladenWeightOverride) {
      const { currentValue, previousValue } = vehicleTechRecord;

      const fieldsChanged = [
        'techRecord_seatsUpperDeck',
        'techRecord_seatsLowerDeck',
        'techRecord_manufactureYear',
        'techRecord_grossKerbWeight'
      ].some(field => currentValue[field] !== previousValue[field]);

      if (fieldsChanged) {
        const newGrossLadenWeight = this.calculateGrossLadenWeight();

        this.form.patchValue(
          {
            techRecord_grossLadenWeight: newGrossLadenWeight
          },
          { emitEvent: false }
        );
      }
    }
  }

  ngOnDestroy(): void {
    this._formSubscription.unsubscribe();
  }

  get template(): FormNode {
    switch (this.vehicleTechRecord.techRecord_vehicleType) {
      case VehicleTypes.PSV:
        return PsvWeightsTemplate;
      case VehicleTypes.HGV:
        return HgvWeight;
      case VehicleTypes.TRL:
        return TrlWeight;
      default:
        throw Error('Incorrect vehicle type!');
    }
  }

  get isPsv(): boolean {
    return this.vehicleTechRecord.techRecord_vehicleType === VehicleTypes.PSV;
  }

  get isHgv(): boolean {
    return this.vehicleTechRecord.techRecord_vehicleType === VehicleTypes.HGV;
  }

  get isTrl(): boolean {
    return this.vehicleTechRecord.techRecord_vehicleType === VehicleTypes.TRL;
  }

  get requiredPlates(): boolean {
    return this.vehicleTechRecord.techRecord_vehicleType !== VehicleTypes.PSV && this.isEditing;
  }

  get types(): typeof FormNodeEditTypes {
    return FormNodeEditTypes;
  }

  get axles(): CustomFormArray {
    return this.form.get(['techRecord_axles']) as CustomFormArray;
  }

  getAxleForm(i: number): CustomFormGroup {
    return this.axles.get([i]) as CustomFormGroup;
  }

  addAxle(): void {
    if (!this.vehicleTechRecord.techRecord_axles || this.vehicleTechRecord.techRecord_axles!.length < 10) {
      this.isError = false;
      this.store.dispatch(addAxle());
    } else {
      this.isError = true;
      this.errorMessage = `Cannot have more than ${10} axles`;
    }
  }

  removeAxle(index: number): void {
    const minLength = this.isTrl ? 1 : 2;

    if (this.vehicleTechRecord.techRecord_axles!.length > minLength) {
      this.isError = false;
      this.store.dispatch(removeAxle({ index }));
    } else {
      this.isError = true;
      this.errorMessage = `Cannot have less than ${minLength} axles`;
    }
  }

  calculateGrossLadenWeight(): number {
    const psvRecord = this.vehicleTechRecord as TechRecordType<'psv'>;
    const techRecord_seatsUpperDeck = psvRecord?.techRecord_seatsUpperDeck ?? 0;
    const techRecord_seatsLowerDeck = psvRecord?.techRecord_seatsLowerDeck ?? 0;
    const techRecord_manufactureYear = psvRecord?.techRecord_manufactureYear ?? 0;
    const techRecord_grossKerbWeight = psvRecord?.techRecord_grossKerbWeight ?? 0;
    const kgAllowedPerPerson = techRecord_manufactureYear >= 1988 ? 65 : 63.5;

    const totalPassengers = techRecord_seatsUpperDeck + techRecord_seatsLowerDeck + 1; // Add 1 for the driver
    return totalPassengers * kgAllowedPerPerson + techRecord_grossKerbWeight;
  }
}
