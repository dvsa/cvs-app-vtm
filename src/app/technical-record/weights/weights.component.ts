import { Component, OnInit, ChangeDetectionStrategy, Input } from '@angular/core';
import { TechRecord } from '../../models/tech-record.model';

@Component({
  selector: 'vtm-weights',
  templateUrl: './weights.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class WeightsComponent implements OnInit {
  @Input() activeRecord: TechRecord;
  @Input() editState: boolean;

  constructor() {}

  ngOnInit() {}
}
