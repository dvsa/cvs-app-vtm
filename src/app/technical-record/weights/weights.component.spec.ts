import {async, ComponentFixture, getTestBed, TestBed} from '@angular/core/testing';
import {TechRecordHelpersService} from '@app/technical-record/tech-record-helpers.service';
import {SharedModule} from '@app/shared/shared.module';
import {CUSTOM_ELEMENTS_SCHEMA} from '@angular/core';
import { WeightsComponent } from './weights.component';

describe('WeightsComponent', () => {

  let component: WeightsComponent;
  let fixture: ComponentFixture<WeightsComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [SharedModule],
      declarations: [
        WeightsComponent,
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(WeightsComponent);
    component = fixture.componentInstance;
    component.activeRecord = {
      'vin': 'XMGDE02FS0H012345',
      'vehicleSize': 'small',
      'testStationName': 'Rowe, Wunsch and Wisoky',
      'vehicleId': 'JY58FPP',
      'vehicleType': 'hgv',
      'grossGbWeight': 3,
      'grossDesignWeight': 2,
      'grossEecWeight': 3,
      'trainGbWeight': 900,
      'trainDesignWeight': 900,
      'trainEecWeight': 400,
      'maxTrainGbWeight': 500,
      'maxTrainDesignWeight': 500,
      'maxTrainEecWeight': 500,
      'axles': [
        {
          'axleNumber': 1,
          'weights': {
            'gbWeight': 500,
            'eecWeight': 500,
            'designWeight': 500
          }
        },
        {
          'axleNumber': 2,
          'weights': {
            'gbWeight': 500,
            'eecWeight': 500,
            'designWeight': 500
          }
        }
      ]
    };
  });

  it('should create view only with populated data', () => {
    fixture.detectChanges();

    expect(component).toBeDefined();
    expect(fixture).toMatchSnapshot();
  });

});
