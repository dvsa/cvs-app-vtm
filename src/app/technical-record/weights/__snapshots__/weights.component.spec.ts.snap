// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeightsComponent should create hgv view only with populated data 1`] = `
<vtm-weights
  activeRecord={[Function Object]}
  techRecHelpers={[Function TechRecordHelpersService]}
>
  <table
    class="govuk-table table-details"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        />
        <th
          class="govuk-table__header govuk-!-width-one-quarter"
          scope="col"
        >
          GB weight (kg)
        </th>
        <th
          class="govuk-table__header govuk-!-width-one-quarter"
          scope="col"
        >
          EEC weight (kg)
        </th>
        <th
          class="govuk-table__header govuk-!-width-one-quarter"
          scope="col"
        >
          Design weight (kg)
        </th>
        <th
          class="govuk-table__cell"
        />
        <th
          class="govuk-table__cell"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-!-font-weight-bold"
        >
          Gross
        </th>
        <td
          class="govuk-table__cell"
          id="test-grossGbWeight"
        >
           3 
        </td>
        <td
          class="govuk-table__cell"
          id="test-grossEecWeight"
        >
           3 
        </td>
        <td
          class="govuk-table__cell"
          id="test-grossDesignWeight"
        >
           2 
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
      </tr>
      
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-!-font-weight-bold"
        >
          Train
        </th>
        <td
          class="govuk-table__cell"
          id="test-trainGbWeight"
        >
           2 
        </td>
        <td
          class="govuk-table__cell"
          id="test-trainDesignWeight"
        >
           - 
        </td>
        <td
          class="govuk-table__cell"
          id="test-trainDesignWeight"
        >
           20 
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
      </tr>
      
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-!-font-weight-bold"
        >
          Max train
        </th>
        <td
          class="govuk-table__cell"
          id="test-maxTrainGbWeight"
        >
           - 
        </td>
        <td
          class="govuk-table__cell"
          id="test-minTrainGbWeight"
        >
           - 
        </td>
        <td
          class="govuk-table__cell"
          id="test-minTrainGbWeight"
        >
           - 
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
      </tr>
      
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-!-font-weight-bold"
        >
          Axle 1
        </th>
        <td
          class="govuk-table__cell"
          id="test-gbWeight-1"
        >
           100
        </td>
        <td
          class="govuk-table__cell"
          id="test-eecWeight-1"
        >
           150
        </td>
        <td
          class="govuk-table__cell"
          id="test-designWeight-1"
        >
           250
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
      </tr>
    </tbody>
  </table>
</vtm-weights>
`;

exports[`WeightsComponent should create trl view only with populated data 1`] = `
<vtm-weights
  activeRecord={[Function Object]}
  techRecHelpers={[Function TechRecordHelpersService]}
>
  <table
    class="govuk-table table-details"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        />
        <th
          class="govuk-table__header govuk-!-width-one-quarter"
          scope="col"
        >
          GB weight (kg)
        </th>
        <th
          class="govuk-table__header govuk-!-width-one-quarter"
          scope="col"
        >
          EEC weight (kg)
        </th>
        <th
          class="govuk-table__header govuk-!-width-one-quarter"
          scope="col"
        >
          Design weight (kg)
        </th>
        <th
          class="govuk-table__cell"
        />
        <th
          class="govuk-table__cell"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-!-font-weight-bold"
        >
          Gross
        </th>
        <td
          class="govuk-table__cell"
          id="test-grossGbWeight"
        >
           3 
        </td>
        <td
          class="govuk-table__cell"
          id="test-grossEecWeight"
        >
           3 
        </td>
        <td
          class="govuk-table__cell"
          id="test-grossDesignWeight"
        >
           2 
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
      </tr>
      
      
      
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__cell govuk-!-font-weight-bold"
        >
          Axle 1
        </th>
        <td
          class="govuk-table__cell"
          id="test-gbWeight-1"
        >
           100
        </td>
        <td
          class="govuk-table__cell"
          id="test-eecWeight-1"
        >
           150
        </td>
        <td
          class="govuk-table__cell"
          id="test-designWeight-1"
        >
           250
        </td>
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        />
      </tr>
    </tbody>
  </table>
</vtm-weights>
`;
